File: src/main/java/site/archive/exception/BaseException.java
package site.archive.exception;

import java.util.Optional;

public class BaseException extends RuntimeException {

    private final ExceptionCode exceptionCode;
    private String additionalMessage;

    public BaseException(ExceptionCode exceptionCode) {
        super(exceptionCode.getMessage());
        this.exceptionCode = exceptionCode;
    }

    public BaseException(String additionalMessage, ExceptionCode exceptionCode) {
        super(String.format("%s : %s", exceptionCode.getMessage(), additionalMessage));
        this.exceptionCode = exceptionCode;
        this.additionalMessage = additionalMessage;
    }

    public ExceptionCode getErrorCode() {
        return exceptionCode;
    }

    public Optional<String> getAdditionalMessage() {
        return Optional.ofNullable(additionalMessage);
    }

}

File: archive-common/src/main/kotlin/site/archive/common/exception/BaseException.kt
package site.archive.common.exception

import java.util.Optional

open class BaseException : RuntimeException {

    val exceptionCode: ExceptionCode
    private var additionalMessage: String? = null

    constructor(exceptionCode: ExceptionCode) : super(exceptionCode.message) {
        this.exceptionCode = exceptionCode
    }

    constructor(additionalMessage: String?, exceptionCode: ExceptionCode) : super("${exceptionCode.message} : $additionalMessage") {
        this.exceptionCode = exceptionCode
        this.additionalMessage = additionalMessage
    }

    fun getAdditionalMessage(): Optional<String> {
        return Optional.ofNullable(additionalMessage)
    }

}